src/Core/src/HttpClient/RetryResponse.php

```diff
@@ -0,0 +1,208 @@
+<?php
+
+namespace AsyncAws\Core\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Chunk\ErrorChunk;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @internal
+ */
+class RetryResponse implements ResponseInterface
+{
+    private $client;
+
+    private $method;
+
+    private $url;
+
+    private $options;
+
+    private $maxTryCount;
+
+    private $logger;
+
+    private $tryCount = 1;
+
+    private $initialized = false;
+
+    /**
+     * @var ResponseInterface
+     */
+    private $inner;
+
+    public function __construct(HttpClientInterface $client, string $method, string $url, array $options, LoggerInterface $logger = null, int $maxTryCount = 3)
+    {
+        $this->client = $client;
+        $this->method = $method;
+        $this->url = $url;
+        $this->options = $options;
+        $this->maxTryCount = $maxTryCount;
+
+        $this->logger = $logger ?? new NullLogger();
+        $this->inner = $this->client->request($this->method, $this->url, $this->options);
+    }
+
+    public function getStatusCode(): int
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getStatusCode();
+    }
+
+    public function getHeaders(bool $throw = true): array
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getHeaders($throw);
+    }
+
+    public function getContent(bool $throw = true): string
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getContent($throw);
+    }
+
+    public function toArray(bool $throw = true): array
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->toArray($throw);
+    }
+
+    public function cancel(): void
+    {
+        $this->initialized = true;
+        $this->inner->cancel();
+    }
+
+    public function getInfo(string $type = null)
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getInfo($type);
+    }
+
+    public static function stream(iterable $responses, float $timeout = null): \Generator
+    {
+        $wrappedResponses = [];
+        $map = new \SplObjectStorage();
+        $client = null;
+
+        foreach ($responses as $r) {
+            if (!$r instanceof self) {
+                throw new \TypeError(sprintf('"%s::stream()" expects parameter 1 to be an iterable of AsyncResponse objects, "%s" given.', static::class, get_debug_type($r)));
```

Comment:
“AsyncResponse”, isn’t that wrong? You mean RetryResponse. 

Let’s use self::class

Reply:
fixed